File: src/utils/voyageCalculations.ts
import { VoyagePlan } from '../types/maritime';

export interface VoyageScenarioResult {
  eta: string;
  fuel: number;
  cost: number;
  waiting: number;
  co2: number;
}

export function scenario(
  voyage: VoyagePlan,
  speed: number,
  bunker: number,
  weather: number,
  berthDelay: number,
  load: number,
  efficiency: number
): VoyageScenarioResult {
  const safeSpeed = Math.max(speed, 1.0);
  const sailingHours = (voyage.remaining_distance_nm / safeSpeed) * weather;
  const currentSpeed = Math.max(voyage.current_speed_knots, 1.0);
  const baselineHours = Math.max(voyage.remaining_distance_nm / currentSpeed, 1.0);
  const baselineRate = voyage.predicted_fuel_tonnes / baselineHours;

  let fuel =
    baselineRate * sailingHours * Math.pow(safeSpeed / currentSpeed, 2) * load * efficiency;
  fuel = Math.max(fuel, 0);

  const waiting = Math.max(0.0, berthDelay);

  // Reference base: "2026-07-23T08:00:00Z"
  const baseTime = new Date('2026-07-23T08:00:00Z').getTime();
  const totalMs = (sailingHours + waiting) * 3600 * 1000;
  const etaDate = new Date(baseTime + totalMs);

  return {
    eta: etaDate.toISOString(),
    fuel,
    cost: fuel * bunker,
    waiting,
    co2: fuel * 3.114,
  };
}

export function voyageKpis(voyages: VoyagePlan[]): Record<string, string | number> {
  const delayed = voyages.filter((v) => new Date(v.predicted_eta) > new Date(v.planned_eta)).length;

  const savings = voyages.map((v) => Math.max(v.planned_fuel_tonnes - v.predicted_fuel_tonnes, 0));
  const totalSaving = savings.reduce((acc, s) => acc + s, 0);

  const costs = voyages.map((v, i) => savings[i] * v.bunker_price_usd_tonne);
  const totalCost = costs.reduce((acc, c) => acc + c, 0);

  const totalWaiting = voyages.reduce((acc, v) => acc + (v.estimated_waiting_hours || 0), 0);
  const co2Reduction = totalSaving * 3.114;

  return {
    'Active voyages': voyages.length,
    'Delayed voyages': delayed,
    'Fuel-saving opportunity': `${totalSaving.toFixed(1)} t`,
    'Estimated cost saving': `$${Math.round(totalCost).toLocaleString()}`,
    'Avoidable waiting time': `${totalWaiting.toFixed(1)} h`,
    'Estimated CO₂ reduction': `${co2Reduction.toFixed(1)} t`,
  };
}

File: legacy/streamlit/utils/voyage_calculations.py
"""Illustrative voyage and fuel scenario calculations."""

from __future__ import annotations

import pandas as pd


def scenario(voyage: pd.Series, speed: float, bunker: float, weather: float, berth_delay: float, load: float, efficiency: float) -> dict[str, float | pd.Timestamp]:
    speed = max(speed, 1.0)
    sailing_hours = voyage["remaining_distance_nm"] / speed * weather
    baseline_rate = voyage["predicted_fuel_tonnes"] / max(voyage["remaining_distance_nm"] / voyage["current_speed_knots"], 1)
    fuel = baseline_rate * sailing_hours * (speed / voyage["current_speed_knots"]) ** 2 * load * efficiency
    fuel = max(fuel, 0)
    waiting = max(0.0, berth_delay)
    eta = pd.Timestamp("2026-07-23 08:00") + pd.to_timedelta(sailing_hours + waiting, unit="h")
    return {
        "eta": eta, "fuel": fuel, "cost": fuel * bunker, "waiting": waiting,
        "co2": fuel * 3.114,
    }


def voyage_kpis(voyages: pd.DataFrame) -> dict[str, str | int]:
    saving = (voyages["planned_fuel_tonnes"] - voyages["predicted_fuel_tonnes"]).clip(lower=0)
    cost = saving * voyages["bunker_price_usd_tonne"]
    return {
        "Active voyages": len(voyages),
        "Delayed voyages": int((voyages["predicted_eta"] > voyages["planned_eta"]).sum()),
        "Fuel-saving opportunity": f"{saving.sum():.1f} t",
        "Estimated cost saving": f"${cost.sum():,.0f}",
        "Avoidable waiting time": f"{voyages['estimated_waiting_hours'].sum():.1f} h",
        "Estimated CO₂ reduction": f"{(saving.sum() * 3.114):.1f} t",
    }

